crates/zksync/core/src/vm/storage_view.rs

```diff
@@ -110,7 +117,9 @@ impl<S: ReadStorage + fmt::Debug> ReadStorage for StorageView<S> {
     }
 
     fn load_factory_dep(&mut self, hash: H256) -> Option<Vec<u8>> {
-        self.storage_handle.load_factory_dep(hash)
+        self.storage_handle
+            .load_factory_dep(hash)
+            .or_else(|| self.persisted_factory_deps.get(&hash).cloned())
```

Comment:
I believe we should check the `persisted_factory_deps` first. The reason being here we're simulating the storage having this value _already_. 

I don't think anything bad can happen with this order as well, but only because we have context. In other parts of the code if the `factory_dep` isn't found then it is fetched via RPC, and given that `storage_handle` is generic over `S`, such a behavior can also be potentially triggered here.

As such, let's check the persisted values before we ask the underlying implementation to fetch it (howsoever it pleases).

Reply:
What I think is that the persisted deps aren't ever cleaned, so if one of the persistent deps is "obsolete" because it's not deployed, it would be fetched from the persisted instead of storage. I can look at how era handles this lookup